metashare/api/sf_run_flow.py

```diff
@@ -46,19 +48,30 @@ def capitalize(s):
     return s[0].upper() + s[1:]
 
 
-def refresh_access_token(*, config, org_name):
+def refresh_access_token(*, config, org_name, scratch_org):
     """
     Construct a new OrgConfig because ScratchOrgConfig tries to use sfdx
     which we don't want now -- this is a total hack which I'll try to
     smooth over with some improvements in CumulusCI
     """
-    org_config = OrgConfig(config, org_name)
-    org_config.refresh_oauth_token = Mock()
-    info = jwt_session(
-        SF_CLIENT_ID, SF_CLIENT_KEY, org_config.username, org_config.instance_url
-    )
-    org_config.config["access_token"] = info["access_token"]
-    return org_config
+    try:
+        org_config = OrgConfig(config, org_name)
+        org_config.refresh_oauth_token = Mock()
+        info = jwt_session(
+            SF_CLIENT_ID, SF_CLIENT_KEY, org_config.username, org_config.instance_url
+        )
+        org_config.config["access_token"] = info["access_token"]
+        return org_config
+    except HTTPError as err:
+        additional_context = ". Are you certain that the org still exists?"
+
+        if get_current_job():
+            job_id = get_current_job().id
+            additional_context += f" If you need support, your job ID is {job_id}."
+
+        err = err.__class__(err.args[0] + additional_context, *err.args[1:],)
+        scratch_org.remove_scratch_org(err)
+        raise err
```

Comment:
@wlonk We have a bit of a possible race condition here that I'm not thrilled about, and I'm trying to think through the best approach. So right now, this fires a `SCRATCH_ORG_REMOVE` WS event with a message, and the front end displays the message and deletes the org accordingly. It also unsubscribes from future WS events for that org.

But then right after that, because this err is raised, the back end also fires the usual error WS event for whatever action was being attempted (e.g. `SCRATCH_ORG_FETCH_CHANGES_FAILED` or `SCRATCH_ORG_COMMIT_CHANGES_FAILED` or whatnot). And if that message is actually received by the front end (that is, if the unsubscription didn't take effect yet), then the following would happen:

1. An additional error message is displayed (not a big deal, and technically accurate anyway).
2. The org is then re-created on the front end, because the additional error message came with a serialized version of the model which is put back into the store.

I could think of a couple of different approaches:

1. Ignore it. I haven't been able to reproduce this locally in any case -- it's more of a theoretical "this seems fragile" race condition.
2. Adjust the front end so that it only _updates_ models in the store, and doesn't insert them if they're not already found.
3. Adjust the model somehow so that the front end can tell from the serialized message that the org has already been deleted on the back end.
4. Some other back end solution...

After walking through those, I think I'm leaning toward #1 or #2? Curious what you think.

Reply:
I've fiddled around with it and thought about it, and I think that 1, with 2 as a mitigation plan fallback should 1 not suffice, is the way to go. We'll want to track if the error is observed in the wild so we know whether we need to switch to 2.